R/utilities-simulation.R

```diff
@@ -244,17 +236,94 @@ createSimulationBatch <- function(simulation, parametersOrPaths = NULL, molecule
+#' moleculesOrPaths = molecules)
+#' simulationBatch2 <- createSimulationBatch(simulation = sim,
+#' parametersOrPaths = parameters,
+#' moleculesOrPaths = molecules)
+#' #Ids of run values
+#' ids <- c()
+#' ids[[1]] <- simulationBatch1$addRunValues(parameterValues = c(1, 2), initialValues = 1)
+#' ids[[2]] <- simulationBatch1$addRunValues(parameterValues = c(1.6, 2.4), initialValues = 3)
+#' ids[[3]] <- simulationBatch2$addRunValues(parameterValues = c(4, 2), initialValues = 4)
+#' ids[[4]] <- simulationBatch2$addRunValues(parameterValues = c(2.6, 4.4), initialValues = 5)
+#' res <- runSimulationBatches(simulationBatches = list(simulationBatch1, simulationBatch2))
+#' }
+runSimulationBatches <- function(simulationBatches, simulationRunOptions = NULL, silentMode = FALSE) {
+  validateIsOfType(simulationBatches, "SimulationBatch")
+  simulationRunner <- getNetTask("ConcurrentSimulationRunner")
+  if (!is.null(simulationRunOptions)) {
+    validateIsOfType(simulationRunOptions, SimulationRunOptions)
+    rClr::clrSet(simulationRunner, "SimulationRunOptions", simulationRunOptions$ref)
+  }
+
+  simulationBatches <- c(simulationBatches)
+  # Result Id <-> simulation batch pointer id map to get the correct simulation for the results.
+  # Using the Id of the pointer instead of the Id of the simulation as multiple
+  # SimulationBatches can be created with the same simulation
+  # Each SimulationBatchRunValues has its own id, which will be the id of the result
+  resultsIdSimulationIdMap <- list()
+  # Map of simulations ids to simulations objects
+  simulationIdSimulationMap <- vector("list", length(simulationBatches))
+  # Iterate through all simulation batches
+  for (simBatchIdx in seq_along(simulationBatches)) {
+    simBatch <- simulationBatches[[simBatchIdx]]
+    # Put the simulation into sim pointer id <-> sim map
+    simId <- deparse(rClr::clrGetExtPtr(simBatch$ref))
+    simulationIdSimulationMap[[simBatchIdx]] <- simBatch$simulation
+    names(simulationIdSimulationMap)[[simBatchIdx]] <- simId
+    # Ids of the values of the batch
+    valuesIds <- simBatch$runValuesIds
+    # All results of this batch have the id of the same simulation
+    resultsIdSimulationIdMap[valuesIds] <- simId
+    # Add the batch to concurrent runner
+    rClr::clrCall(simulationRunner, "AddSimulationBatchOption", simBatch$ref)
+  }
+
+  # Run the batch with the ConcurrentSimulationRunner
+  results <- rClr::clrCall(simulationRunner, "RunConcurrently")
+  simulationResults <- .getConcurrentSimulationRunnerResults(results = results, resultsIdSimulationIdMap = resultsIdSimulationIdMap, simulationIdSimulationMap = simulationIdSimulationMap, silentMode = silentMode)
+
+  # output: list of lists of SimulationResults, one list per SimulationBatch
+  output <- lapply(names(simulationIdSimulationMap), function(simId) {
+    simulationResults[which(resultsIdSimulationIdMap == simId)]
+  })
+  # Not sure this is required
+  rClr::clrCall(simulationRunner, "Dispose")

```

Comment:
there is an issue here. If this object is returned via a Task, it might be singleton and should NOT Be disposed.

Reply:
Ok, I don't have experience with this. Removing the dispose.

Reply:
we might need this. A task should not be disposable. It might be something I need to discuss with @abdelr

Reply:
in short,if this is disposable, we should get a new instance everytime.

Reply:
Ok I checked. Dispose is fine and should be call in that case explicitly. Not a big deal if not but it will release stuff immediately.